Receive and save the script to disk

WilliamMcCumstie opened the pull request:
Previously only the filepath to the script was sent to the server, as it is assumed all the nodes share a filesystem. This assumption maybe invalid and has been removed.

Instead the script must be sent with the job when it is created. It is than saved to disk so it can be retrieved latter. This required various changes to the internals to remove the old `script` method and replace it with `read_script`. The over-arching design principle is the `script` should not be cached on a `Job` object as it will remain in ram. Instead it has been written to disk asap and the read from disk when required.

The work flow for when the script is deleted could be improved. ATM it is deleted by scheduler

Finally because we are working with files, it is handy to fake the filesystem to prevent the excessive build up of files. This needs to be activated manually to prevent any surprises.

PS Validation errors now return 422 instead of 500

WilliamMcCumstie committed:
Do not serialize the script with the Job

The `script` key will no longer be the path to the job script,
but the body of the script itself

As such it should not be returned with the Job as it maybe quite large

WilliamMcCumstie committed:
Define where the job scripts will be stored

WilliamMcCumstie committed:
Explicitly require the create route to add a job

The `validate!` method is ran on various different routes (such as
`update`). This isn't an issue yet as only the `create` route exists.

However a `@created` check should still be preformed

WilliamMcCumstie committed:
Upgrade the Job.script to Job.read_script

This commit is primarily concerned with changing the `script` method
to `read_script`

The validation has been removed so it can be handled differently

TBA The `script` will be the entire text body and not the file path. To
facilitate this, the script should not be stored on the Job object but
instead read from it

WilliamMcCumstie committed:
Setup a demo app for future controller specs

WilliamMcCumstie committed:
Write the script to disk when creating a job

The script is no longer stored in ram but instead on disk

WilliamMcCumstie committed:
Remove the script file at the end of the jobs lifecycle

This happens when it is removed from its scheduler, which isn't
the ideal place but is okay for now

WilliamMcCumstie committed:
Ensure the create errors if the script isn't provided

This is done via a virtual attribute 'script_provided' which avoids
caching the entire script in ram

It is done via a validation so the error mode is consistent

WilliamMcCumstie committed:
Add the travis config

WilliamMcCumstie commented on lib/flight_scheduler/event_processor.rb:
```diff
@@ -50,6 +50,7 @@ def cancel_job(job)
     Async.logger.info("Canceling job #{job.id}")
     return unless %w(PENDING RUNNING).include?(job.state)
 
+    # NOTE: Is the above and below state lines correct? Needs review
```
I came across this section whilst refactoring but couldn't determine what it was meant to do. Two things:

1. Is line 54 meant to be `job.state = CANCELLED`? ATM it is doing an equality check but not using the result.
2. What is the purpose of line 55 `return if <pending>` considering we do a similar check above?

benarmston replied:
Job cancellation has seen some refactoring in #8.  FWIW: line 54 should have been an assignment.  Line 51, prevents cancelling completed jobs and line 55 prevents attempting to kill processes for pending jobs (obviously the processes don't exist).

WilliamMcCumstie replied:
Removed the comment in light of your changes:
https://github.com/openflighthpc/flight-scheduler-controller/pull/7/commits/06e21a3335ad12454e58b36fdb028b60f5378fd9

benarmston commented on app.rb:
```diff
@@ -190,8 +195,11 @@ def find(id)
 
       def validate!
-        if resource.validate!
+        if @created && resource.validate!
```
What's the purpose of `@created` here?  I thought `validate!` was only called after the `create` action.  Is it being called at other times too?

WilliamMcCumstie replied:
ATM you are correct that `validate!` only gets called during `create`, however in general this is not the case. Adding a `update` (possible `pluck`, `prune`, `graft`, etc) route will also trigger the `validate!` and break the scheduler.

By using a `@create` variable it explicitly prevents jobs being re-added in the future.

tl;dr It is not required beyond best practice

benarmston replied:
Fair enough.

benarmston commented on lib/flight_scheduler/schedulers/fifo_scheduler.rb:
```diff
@@ -42,8 +42,13 @@ def add_job(job)
+  # consider refactoring
   def remove_job(job)
     @queue.delete(job)
+    job.cleanup
```
This is fine for now.  But removing a job from the scheduler and cleaning up a job are two different concerns.  Once this branch and the array tasks branch are both merged, we can look at a better place for this to live.

WilliamMcCumstie replied:
:+1:

benarmston reviewed (commented):
I've left a comment in-line below.

WilliamMcCumstie committed:
Remove comment as the issue will be resolved

The relevant lines are code are being refactored making
the comment obsolete

WilliamMcCumstie committed:
Require the script-name to be submitted

This isn't used to fulfil the request but instead to be a human
identifiable label

This means it still needs to be set with the request

benarmston commented:
LGTM.
